File: sdk/factory.ts
import { ccc } from "@ckb-ccc/core";

export const FACTORY_DATA_SIZE = 256;
export const DEX_INSTANCE_SIZE = 192;
export const POOL_DATA_SIZE = 152;
export const DEFAULT_FACTORY_FEE_BPS = 500;
export const MIN_DEX_FEE_BPS = 10;
export const MAX_DEX_FEE_BPS = 500;
export const CREATOR_FEE_BPS = 300;
export const DEFAULT_CREATION_FEE_CKB = 5000n;

export interface FactoryData {
  ownerLockHash: string;
  factoryFeeBps: number;
  dexCount: bigint;
  totalFeesCollected: bigint;
  minimumDexFeeBps: number;
  maximumDexFeeBps: number;
  creationFeeCkb: bigint;
  totalCreationFees: bigint;
  bump: bigint;
}

export interface DexInstanceData {
  dexId: string;
  dexNameHash: string;
  ownerLockHash: string;
  dexFeeBps: number;
  factoryFeeBps: number;
  creatorFeeBps: number;
  lpFeeBps: number;
  poolCount: bigint;
  totalVolume: bigint;
  totalFeesToFactory: bigint;
  createdAt: bigint;
  bump: bigint;
}

export interface PoolData {
  poolId: string;
  tokenATypeHash: string;
  tokenBTypeHash: string;
  reserveA: bigint;
  reserveB: bigint;
  feeBps: bigint;
  lpSupply: bigint;
  kLast: bigint;
  bump: bigint;
  createdAt: bigint;
}

export interface CreateDexParams {
  dexName: string;
  dexFeeBps: number;
  ownerAddress: string;
  tokenA: string;
  tokenB: string;
  initialLiquidityA: bigint;
  initialLiquidityB: bigint;
}

export interface FeeBreakdown {
  dexFeeBps: number;
  factoryFeeBps: number;
  creatorFeeBps: number;
  lpFeeBps: number;
  factoryFeePercent: number;
  creatorFeePercent: number;
}

export interface DexInstance extends DexInstanceData {
  dexName?: string;
  metadata?: DexMetadata;
}

export interface DexMetadata {
  name: string;
  description: string;
  website?: string;
  logoUrl?: string;
  twitter?: string;
  discord?: string;
}

export interface FactoryConfig {
  factoryFeeBps: number;
  minDexFeeBps: number;
  maxDexFeeBps: number;
  creationFeeCkb: bigint;
}

export interface InitializeFactoryParams {
  ownerLockHash: string;
  factoryFeeBps?: number;
  minDexFeeBps?: number;
  maxDexFeeBps?: number;
  creationFeeCkb?: bigint;
}

export function encodeFactoryData(data: FactoryData): Uint8Array {
  const bytes = new Uint8Array(FACTORY_DATA_SIZE);
  const ownerBytes = ccc.bytesFrom(ccc.hexFrom(data.ownerLockHash));
  bytes.set(ownerBytes.slice(0, 32), 0);
  bytes.set(ccc.numLeToBytes(data.factoryFeeBps, 2), 40);
  bytes.set(ccc.numLeToBytes(data.dexCount, 8), 48);
  bytes.set(ccc.numLeToBytes(data.totalFeesCollected, 8), 56);
  bytes.set(ccc.numLeToBytes(data.minimumDexFeeBps, 2), 64);
  bytes.set(ccc.numLeToBytes(data.maximumDexFeeBps, 2), 72);
  bytes.set(ccc.numLeToBytes(data.creationFeeCkb, 8), 80);
  bytes.set(ccc.numLeToBytes(data.totalCreationFees, 8), 88);
  bytes.set(ccc.numLeToBytes(data.bump, 8), 96);
  return bytes;
}

export function decodeFactoryData(bytes: Uint8Array): FactoryData {
  if (bytes.length !== FACTORY_DATA_SIZE) {
    throw new Error(
      `Invalid factory data length: expected ${FACTORY_DATA_SIZE}, got ${bytes.length}`,
    );
  }
  return {
    ownerLockHash: ccc.hexFrom(bytes.slice(0, 32)),
    factoryFeeBps: ccc.numLeFromBytes(bytes.slice(40, 42)),
    dexCount: ccc.numLeFromBytes(bytes.slice(48, 56)),
    totalFeesCollected: ccc.numLeFromBytes(bytes.slice(56, 64)),
    minimumDexFeeBps: ccc.numLeFromBytes(bytes.slice(64, 66)),
    maximumDexFeeBps: ccc.numLeFromBytes(bytes.slice(72, 74)),
    creationFeeCkb: ccc.numLeFromBytes(bytes.slice(80, 88)),
    totalCreationFees: ccc.numLeFromBytes(bytes.slice(88, 96)),
    bump: ccc.numLeFromBytes(bytes.slice(96, 104)),
  };
}

export function encodeDexInstanceData(data: DexInstanceData): Uint8Array {
  const bytes = new Uint8Array(DEX_INSTANCE_SIZE);
  const dexIdBytes = ccc.bytesFrom(ccc.hexFrom(data.dexId));
  bytes.set(dexIdBytes.slice(0, 32), 0);
  const nameHashBytes = ccc.bytesFrom(ccc.hexFrom(data.dexNameHash));
  bytes.set(nameHashBytes.slice(0, 32), 32);
  const ownerBytes = ccc.bytesFrom(ccc.hexFrom(data.ownerLockHash));
  bytes.set(ownerBytes.slice(0, 32), 64);
  bytes.set(ccc.numLeToBytes(data.dexFeeBps, 2), 96);
  bytes.set(ccc.numLeToBytes(data.factoryFeeBps, 2), 104);
  bytes.set(ccc.numLeToBytes(data.creatorFeeBps, 2), 112);
  bytes.set(ccc.numLeToBytes(data.lpFeeBps, 2), 120);
  bytes.set(ccc.numLeToBytes(data.poolCount, 8), 128);
  bytes.set(ccc.numLeToBytes(data.totalVolume, 8), 136);
  bytes.set(ccc.numLeToBytes(data.totalFeesToFactory, 8), 144);
  bytes.set(ccc.numLeToBytes(data.createdAt, 8), 152);
  bytes.set(ccc.numLeToBytes(data.bump, 8), 160);
  return bytes;
}

export function decodeDexInstanceData(bytes: Uint8Array): DexInstanceData {
  if (bytes.length !== DEX_INSTANCE_SIZE) {
    throw new Error(
      `Invalid DEX instance data length: expected ${DEX_INSTANCE_SIZE}, got ${bytes.length}`,
    );
  }
  return {
    dexId: ccc.hexFrom(bytes.slice(0, 32)),
    dexNameHash: ccc.hexFrom(bytes.slice(32, 64)),
    ownerLockHash: ccc.hexFrom(bytes.slice(64, 96)),
    dexFeeBps: ccc.numLeFromBytes(bytes.slice(96, 98)),
    factoryFeeBps: ccc.numLeFromBytes(bytes.slice(104, 106)),
    creatorFeeBps: ccc.numLeFromBytes(bytes.slice(112, 114)),
    lpFeeBps: ccc.numLeFromBytes(bytes.slice(120, 122)),
    poolCount: ccc.numLeFromBytes(bytes.slice(128, 136)),
    totalVolume: ccc.numLeFromBytes(bytes.slice(136, 144)),
    totalFeesToFactory: ccc.numLeFromBytes(bytes.slice(144, 152)),
    createdAt: ccc.numLeFromBytes(bytes.slice(152, 160)),
    bump: ccc.numLeFromBytes(bytes.slice(160, 168)),
  };
}

export function encodePoolData(data: PoolData): Uint8Array {
  const bytes = new Uint8Array(POOL_DATA_SIZE);
  const poolIdBytes = ccc.bytesFrom(ccc.hexFrom(data.poolId));
  bytes.set(poolIdBytes.slice(0, 32), 0);
  const tokenABytes = ccc.bytesFrom(ccc.hexFrom(data.tokenATypeHash));
  bytes.set(tokenABytes.slice(0, 32), 32);
  const tokenBBytes = ccc.bytesFrom(ccc.hexFrom(data.tokenBTypeHash));
  bytes.set(tokenBBytes.slice(0, 32), 64);
  bytes.set(ccc.numLeToBytes(data.reserveA, 8), 64);
  bytes.set(ccc.numLeToBytes(data.reserveB, 8), 72);
  bytes.set(ccc.numLeToBytes(Number(data.feeBps), 2), 80);
  bytes.set(ccc.numLeToBytes(data.lpSupply, 8), 88);
  bytes.set(ccc.numLeToBytes(data.kLast, 8), 96);
  bytes.set(ccc.numLeToBytes(data.bump, 8), 104);
  bytes.set(ccc.numLeToBytes(data.createdAt, 8), 112);
  return bytes;
}

export function decodePoolData(bytes: Uint8Array): PoolData {
  if (bytes.length !== POOL_DATA_SIZE) {
    throw new Error(
      `Invalid pool data length: expected ${POOL_DATA_SIZE}, got ${bytes.length}`,
    );
  }
  return {
    poolId: ccc.hexFrom(bytes.slice(0, 32)),
    tokenATypeHash: ccc.hexFrom(bytes.slice(32, 64)),
    tokenBTypeHash: ccc.hexFrom(bytes.slice(64, 96)),
    reserveA: ccc.numLeFromBytes(bytes.slice(96, 104)),
    reserveB: ccc.numLeFromBytes(bytes.slice(104, 112)),
    feeBps: BigInt(ccc.numLeFromBytes(bytes.slice(112, 114))),
    lpSupply: ccc.numLeFromBytes(bytes.slice(120, 128)),
    kLast: ccc.numLeFromBytes(bytes.slice(128, 136)),
    bump: ccc.numLeFromBytes(bytes.slice(136, 144)),
    createdAt: ccc.numLeFromBytes(bytes.slice(144, 152)),
  };
}

export function calculateFeeBreakdown(
  dexFeeBps: number,
  factoryFeeBps: number,
  creatorFeeBps: number,
): FeeBreakdown {
  const factoryCut = Math.floor((dexFeeBps * factoryFeeBps) / 10000);
  const creatorCut = Math.floor((dexFeeBps * creatorFeeBps) / 10000);
  const lpCut = dexFeeBps - factoryCut - creatorCut;
  return {
    dexFeeBps,
    factoryFeeBps: factoryCut,
    creatorFeeBps: creatorCut,
    lpFeeBps: lpCut,
    factoryFeePercent: dexFeeBps > 0 ? (factoryCut / dexFeeBps) * 100 : 0,
    creatorFeePercent: dexFeeBps > 0 ? (creatorCut / dexFeeBps) * 100 : 0,
  };
}

export function calculateSwapFees(
  volume: bigint,
  dexFeeBps: number,
  factoryFeeBps: number,
): [bigint, bigint] {
  const totalFee = (volume * BigInt(dexFeeBps)) / 10000n;
  const factoryFee = (volume * BigInt(factoryFeeBps)) / 10000n;
  const lpFee = totalFee - factoryFee;
  return [factoryFee, lpFee];
}

export function hashDexName(name: string): string {
  const crypto = require("crypto");
  return (
    "0x" + crypto.createHash("sha256").update(name).digest().toString("hex")
  );
}

export function validateDexName(name: string): boolean {
  if (!name || name.length === 0 || name.length > 32) {
    return false;
  }
  const validPattern = /^[a-zA-Z0-9 _-]+$/;
  return validPattern.test(name);
}

export function generateDexId(
  factoryHash: string,
  ownerHash: string,
  bump: bigint,
): string {
  const factoryBytes = ccc.bytesFrom(ccc.hexFrom(factoryHash));
  const ownerBytes = ccc.bytesFrom(ccc.hexFrom(ownerHash));
  const bumpBytes = ccc.numLeToBytes(bump, 8);
  const id = new Uint8Array(32);
  for (let i = 0; i < 32; i++) {
    id[i] = factoryBytes[i] ^ ownerBytes[i];
    if (i < 8) {
      id[i] ^= bumpBytes[i];
    }
  }
  return ccc.hexFrom(id);
}

export function calculateInitialLP(reserveA: bigint, reserveB: bigint): bigint {
  const product = reserveA * reserveB;
  const sqrt = BigInt(Math.floor(Number(product) ** 0.5));
  const MINIMUM_LIQUIDITY = 1000n;
  return sqrt > MINIMUM_LIQUIDITY ? sqrt - MINIMUM_LIQUIDITY : 0n;
}

export class DexFactoryClient {
  private client: ccc.Client;
  private factoryScriptHash?: string;
  private factoryCell?: { txHash: string; index: number };

  constructor(rpcUrl: string) {
    this.client = new ccc.Client({ url: rpcUrl });
  }

  async connect(): Promise<void> {
    await this.client.getFeeRate();
  }

  setFactoryScriptHash(scriptHash: string): void {
    this.factoryScriptHash = scriptHash;
  }

  setFactoryCell(txHash: string, index: number): void {
    this.factoryCell = { txHash, index };
  }

  async findFactoryCell(): Promise<{
    cell: ccc.Cell;
    data: FactoryData;
  } | null> {
    if (!this.factoryScriptHash) {
      throw new Error("Factory script hash not set");
    }
    const script = new ccc.Script(
      this.factoryScriptHash,
      "0x",
      ccc.hashTypeFrom("type"),
    );
    const cells = await this.client.findCells(
      {
        script,
        scriptType: "type",
        scriptSearchMode: "prefix",
      },
      "asc",
      "desc",
      10,
    );
    if (!cells || cells.length === 0) {
      return null;
    }
    const factoryCell = cells[0];
    const args = factoryCell.cellOutput.type?.args || new Uint8Array(256);
    const data = decodeFactoryData(args);
    return { cell: factoryCell, data };
  }

  async getFactoryData(): Promise<FactoryData> {
    const result = await this.findFactoryCell();
    if (!result) {
      throw new Error("Factory cell not found");
    }
    return result.data;
  }

  async getConfig(): Promise<FactoryConfig> {
    const factory = await this.getFactoryData();
    return {
      factoryFeeBps: factory.factoryFeeBps,
      minDexFeeBps: factory.minimumDexFeeBps,
      maxDexFeeBps: factory.maximumDexFeeBps,
      creationFeeCkb: factory.creationFeeCkb,
    };
  }

  async validateDexFee(feeBps: number): Promise<boolean> {
    const config = await this.getConfig();
    return feeBps >= config.minDexFeeBps && feeBps <= config.maxDexFeeBps;
  }

  async calculateFeeBreakdown(dexFeeBps: number): Promise<FeeBreakdown> {
    const config = await this.getConfig();
    return calculateFeeBreakdown(
      dexFeeBps,
      config.factoryFeeBps,
      CREATOR_FEE_BPS,
    );
  }

  async initializeFactory(
    params: InitializeFactoryParams,
    ownerPrivateKey: string,
  ): Promise<{
    txHash: string;
    factoryAddress: string;
  }> {
    const ownerLockHash = params.ownerLockHash;
    if (!ownerLockHash) {
      throw new Error("Owner lock hash is required");
    }
    const factoryData: FactoryData = {
      ownerLockHash,
      factoryFeeBps: params.factoryFeeBps || DEFAULT_FACTORY_FEE_BPS,
      dexCount: 0n,
      totalFeesCollected: 0n,
      minimumDexFeeBps: params.minDexFeeBps || MIN_DEX_FEE_BPS,
      maximumDexFeeBps: params.maxDexFeeBps || MAX_DEX_FEE_BPS,
      creationFeeCkb: params.creationFeeCkb || DEFAULT_CREATION_FEE_CKB,
      totalCreationFees: 0n,
      bump: 1n,
    };
    const dataBytes = encodeFactoryData(factoryData);
    const dataHex = ccc.hexFrom(dataBytes);
    if (!this.factoryScriptHash) {
      throw new Error("Factory script hash not set");
    }
    const factoryScript = new ccc.Script(
      this.factoryScriptHash,
      dataHex,
      ccc.hashTypeFrom("type"),
    );
    const capacity = 350n * 10n ** 8n;
    const tx = await this.client.buildTransaction({
      outputs: [
        {
          lock: await this._getOwnerLock(ownerPrivateKey),
          type: factoryScript,
          capacity,
        },
      ],
      outputsData: [dataHex],
    });
    const signer = new ccc.SignerCkbPrivateKey(this.client, ownerPrivateKey);
    const signedTx = await signer.signOnlyTransaction(tx);
    const txHash = await this.client.sendTransaction(signedTx);
    const factoryAddress = this._encodeAddress(factoryScript);
    this.factoryCell = { txHash, index: 0 };
    return { txHash, factoryAddress };
  }

  async createDex(
    params: CreateDexParams,
    ownerPrivateKey: string,
  ): Promise<{
    txHash: string;
    dexId: string;
    dexAddress: string;
    poolAddress: string;
  }> {
    if (!validateDexName(params.dexName)) {
      throw new Error("Invalid DEX name");
    }
    if (!(await this.validateDexFee(params.dexFeeBps))) {
      const config = await this.getConfig();
      throw new Error(
        `DEX fee must be between ${config.minDexFeeBps} and ${config.maxDexFeeBps} bps`,
      );
    }
    const config = await this.getConfig();
    const feeBreakdown = calculateFeeBreakdown(
      params.dexFeeBps,
      config.factoryFeeBps,
      CREATOR_FEE_BPS,
    );
    const factoryResult = await this.findFactoryCell();
    if (!factoryResult) {
      throw new Error("Factory cell not found - must initialize first");
    }
    const factoryHash = factoryResult.cell.cellOutput.type?.codeHash || "";
    const bump = BigInt(Date.now());
    const dexId = generateDexId(factoryHash, params.ownerAddress, bump);
    const dexData: DexInstanceData = {
      dexId,
      dexNameHash: hashDexName(params.dexName),
      ownerLockHash: params.ownerAddress,
      dexFeeBps: feeBreakdown.dexFeeBps,
      factoryFeeBps: feeBreakdown.factoryFeeBps,
      creatorFeeBps: feeBreakdown.creatorFeeBps,
      lpFeeBps: feeBreakdown.lpFeeBps,
      poolCount: 0n,
      totalVolume: 0n,
      totalFeesToFactory: 0n,
      createdAt: BigInt(Math.floor(Date.now() / 1000)),
      bump,
    };
    const lpSupply = calculateInitialLP(
      params.initialLiquidityA,
      params.initialLiquidityB,
    );
    const poolData: PoolData = {
      poolId: dexId,
      tokenATypeHash: params.tokenA,
      tokenBTypeHash: params.tokenB,
      reserveA: params.initialLiquidityA,
      reserveB: params.initialLiquidityB,
      feeBps: BigInt(feeBreakdown.dexFeeBps),
      lpSupply,
      kLast: 0n,
      bump,
      createdAt: dexData.createdAt,
    };
    const ownerLock = await this._getOwnerLock(ownerPrivateKey);
    const creationFeeCapacity = config.creationFeeCkb * 10n ** 8n;
    const poolCapacity = 200n * 10n ** 8n;
    const dexCapacity = 250n * 10n ** 8n;
    const lpCapacity = 100n * 10n ** 8n;
    const outputs = [
      {
        lock: ownerLock,
        type: new ccc.Script(
          this.factoryScriptHash!,
          dexId,
          ccc.hashTypeFrom("type"),
        ),
        capacity: dexCapacity,
      },
      {
        lock: ownerLock,
        type: new ccc.Script(
          this._getPoolScriptHash(),
          dexId,
          ccc.hashTypeFrom("type"),
        ),
        capacity: poolCapacity,
      },
      {
        lock: await this._getOwnerLockByHash(factoryResult.data.ownerLockHash),
        capacity: creationFeeCapacity,
      },
      {
        lock: ownerLock,
        type: new ccc.Script(
          this._getLpTokenScriptHash(),
          dexId,
          ccc.hashTypeFrom("type"),
        ),
        capacity: lpCapacity,
      },
    ];
    const outputsData = [
      ccc.hexFrom(encodeDexInstanceData(dexData)),
      ccc.hexFrom(encodePoolData(poolData)),
      "0x",
      ccc.hexFrom(ccc.numLeToBytes(lpSupply, 16)),
    ];
    const tx = await this.client.buildTransaction({
      outputs,
      outputsData,
    });
    const signer = new ccc.SignerCkbPrivateKey(this.client, ownerPrivateKey);
    const signedTx = await signer.signOnlyTransaction(tx);
    const txHash = await this.client.sendTransaction(signedTx);
    const dexAddress = this._encodeAddress(outputs[0].type);
    const poolAddress = this._encodeAddress(outputs[1].type);
    return { txHash, dexId, dexAddress, poolAddress };
  }

  private async _getOwnerLock(privateKey: string): Promise<ccc.Script> {
    const signer = new ccc.SignerCkbPrivateKey(this.client, privateKey);
    const pubKeyBytes = ccc.bytesFrom(signer.publicKey);
    const pubKeyHash = ccc.hashCkb(pubKeyBytes).slice(0, 20);
    return new ccc.Script(
      "0x9bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce8",
      pubKeyHash,
      ccc.hashTypeFrom("type"),
    );
  }

  private async _getOwnerLockByHash(lockHash: string): Promise<ccc.Script> {
    return new ccc.Script(
      "0x9bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce8",
      lockHash,
      ccc.hashTypeFrom("type"),
    );
  }

  private _getPoolScriptHash(): string {
    try {
      const deployment = require("../deployments/devnet-deployments.json");
      return (
        deployment.contracts["pool"]?.typeScript?.codeHash ||
        "0x0000000000000000000000000000000000000000000000000000000000000000"
      );
    } catch {
      return "0x0000000000000000000000000000000000000000000000000000000000000000";
    }
  }

  private _getLpTokenScriptHash(): string {
    try {
      const deployment = require("../deployments/devnet-deployments.json");
      return (
        deployment.contracts["pool"]?.typeScript?.codeHash ||
        "0x0000000000000000000000000000000000000000000000000000000000000000"
      );
    } catch {
      return "0x0000000000000000000000000000000000000000000000000000000000000000";
    }
  }

  private _encodeAddress(script: ccc.Script): string {
    const BECH32M_CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l";
    const BECH32M_CONST = 0x2bc830a3;
    const prefix = "ckt";
    const codeHashBytes = ccc.bytesFrom(ccc.hexFrom(script.codeHash));
    const hashTypeByte =
      script.hashType === "type"
        ? new Uint8Array([0x01])
        : new Uint8Array([0x00]);
    const argsBytes = ccc.bytesFrom(ccc.hexFrom(script.args));
    const payload = new Uint8Array(codeHashBytes.length + 1 + argsBytes.length);
    payload.set(codeHashBytes, 0);
    payload.set(hashTypeByte, 32);
    payload.set(argsBytes, 33);
    const convertBits = (
      data: Uint8Array,
      fromBits: number,
      toBits: number,
      pad = true,
    ): number[] => {
      const ret: number[] = [];
      let acc = 0;
      let bits = 0;
      const maxv = (1 << toBits) - 1;
      for (let i = 0; i < data.length; i++) {
        const value = data[i];
        acc = (acc << fromBits) | value;
        bits += fromBits;
        while (bits >= toBits) {
          bits -= toBits;
          ret.push((acc >> bits) & maxv);
        }
      }
      if (pad && bits > 0) {
        ret.push((acc << (toBits - bits)) & maxv);
      }
      return ret;
    };
    const payload5bit = convertBits(payload, 8, 5, true);
    const polymod = (values: number[]): number => {
      let chk = 1;
      const prefixValues: number[] = [];
      for (let i = 0; i < prefix.length; i++) {
        prefixValues.push(prefix.charCodeAt(i) & 0x1f);
      }
      prefixValues.push(0);
      for (const v of prefixValues) {
        const b = (chk >> 25) & 0x1f;
        chk = ((chk & 0x1ffffff) << 5) ^ v;
        if (b & 1) chk ^= 0x3b6a57b2;
        if (b & 2) chk ^= 0x26508e6d;
        if (b & 4) chk ^= 0x1ea119fa;
        if (b & 8) chk ^= 0x3d4233dd;
        if (b & 16) chk ^= 0x2a1462b3;
      }
      for (const v of values) {
        const b = (chk >> 25) & 0x1f;
        chk = ((chk & 0x1ffffff) << 5) ^ v;
        if (b & 1) chk ^= 0x3b6a57b2;
        if (b & 2) chk ^= 0x26508e6d;
        if (b & 4) chk ^= 0x1ea119fa;
        if (b & 8) chk ^= 0x3d4233dd;
        if (b & 16) chk ^= 0x2a1462b3;
      }
      return chk;
    };
    const checksumInput = [...payload5bit];
    let polymodInput = [...checksumInput];
    for (let i = 0; i < 6; i++) polymodInput.push(0);
    const checksum = polymod(polymodInput) ^ BECH32M_CONST;
    const checksum5bit: number[] = [];
    for (let i = 0; i < 6; i++) {
      checksum5bit.push((checksum >> (5 * (5 - i))) & 0x1f);
    }
    let result = prefix + "1";
    for (const v of payload5bit) result += BECH32M_CHARSET[v];
    for (const v of checksum5bit) result += BECH32M_CHARSET[v];
    return result;
  }
}

File: orhex/contracts/factory/src/factory.rs
use ckb_std::high_level::load_cell_data;
use ckb_std::ckb_constants::Source;

pub const FACTORY_DATA_SIZE: usize = 256;
pub const DEX_INSTANCE_SIZE: usize = 192;
pub const MAX_FACTORY_FEE_BPS: u16 = 1000;
pub const DEFAULT_FACTORY_FEE_BPS: u16 = 500;
pub const DEFAULT_CREATOR_FEE_BPS: u16 = 300;
pub const MIN_DEX_FEE_BPS: u16 = 10;
pub const MAX_DEX_FEE_BPS: u16 = 500;

#[repr(C)]
#[derive(Debug)]
pub struct FactoryData {
    pub owner_lock_hash: [u8; 32],
    pub factory_fee_bps: u16,
    pub dex_count: u64,
    pub total_fees_collected: u64,
    pub minimum_dex_fee_bps: u16,
    pub maximum_dex_fee_bps: u16,
    pub creation_fee_ckb: u64,
    pub total_creation_fees: u64,
    pub creator_fee_bps: u16,
    pub bump: u64,
    pub reserved: [u8; 152],
}

impl FactoryData {
    pub fn new(owner_lock_hash: [u8; 32], factory_fee_bps: u16, creation_fee_ckb: u64, creator_fee_bps: u16) -> Self {
        Self {
            owner_lock_hash,
            factory_fee_bps,
            dex_count: 0,
            total_fees_collected: 0,
            minimum_dex_fee_bps: MIN_DEX_FEE_BPS,
            maximum_dex_fee_bps: MAX_DEX_FEE_BPS,
            creation_fee_ckb,
            total_creation_fees: 0,
            creator_fee_bps,
            bump: 0,
            reserved: [0u8; 152],
        }
    }

    pub fn to_bytes(&self) -> [u8; FACTORY_DATA_SIZE] {
        let mut bytes = [0u8; FACTORY_DATA_SIZE];
        bytes[0..32].copy_from_slice(&self.owner_lock_hash);
        bytes[40..42].copy_from_slice(&self.factory_fee_bps.to_le_bytes());
        bytes[48..56].copy_from_slice(&self.dex_count.to_le_bytes());
        bytes[56..64].copy_from_slice(&self.total_fees_collected.to_le_bytes());
        bytes[64..66].copy_from_slice(&self.minimum_dex_fee_bps.to_le_bytes());
        bytes[72..74].copy_from_slice(&self.maximum_dex_fee_bps.to_le_bytes());
        bytes[80..88].copy_from_slice(&self.creation_fee_ckb.to_le_bytes());
        bytes[88..96].copy_from_slice(&self.total_creation_fees.to_le_bytes());
        bytes[96..98].copy_from_slice(&self.creator_fee_bps.to_le_bytes());
        bytes[104..112].copy_from_slice(&self.bump.to_le_bytes());
        bytes
    }

    pub fn from_bytes(bytes: &[u8]) -> Result<Self, &'static str> {
        if bytes.len() != FACTORY_DATA_SIZE {
            return Err("Invalid factory data length");
        }
        let mut owner_lock_hash = [0u8; 32];
        owner_lock_hash.copy_from_slice(&bytes[0..32]);
        let factory_fee_bps = u16::from_le_bytes(bytes[40..42].try_into().unwrap());
        let dex_count = u64::from_le_bytes(bytes[48..56].try_into().unwrap());
        let total_fees_collected = u64::from_le_bytes(bytes[56..64].try_into().unwrap());
        let minimum_dex_fee_bps = u16::from_le_bytes(bytes[64..66].try_into().unwrap());
        let maximum_dex_fee_bps = u16::from_le_bytes(bytes[72..74].try_into().unwrap());
        let creation_fee_ckb = u64::from_le_bytes(bytes[80..88].try_into().unwrap());
        let total_creation_fees = u64::from_le_bytes(bytes[88..96].try_into().unwrap());
        let creator_fee_bps = u16::from_le_bytes(bytes[96..98].try_into().unwrap());
        let bump = u64::from_le_bytes(bytes[104..112].try_into().unwrap());
        Ok(Self {
            owner_lock_hash, factory_fee_bps, dex_count, total_fees_collected,
            minimum_dex_fee_bps, maximum_dex_fee_bps, creation_fee_ckb,
            total_creation_fees, creator_fee_bps, bump, reserved: [0u8; 152],
        })
    }

    pub fn calculate_factory_fee(&self, dex_fee_bps: u16) -> u16 {
        (dex_fee_bps as u32 * self.factory_fee_bps as u32 / 10000) as u16
    }

    pub fn calculate_creator_fee(&self, dex_fee_bps: u16) -> u16 {
        (dex_fee_bps as u32 * self.creator_fee_bps as u32 / 10000) as u16
    }

    pub fn calculate_lp_fee(&self, dex_fee_bps: u16) -> u16 {
        dex_fee_bps - self.calculate_factory_fee(dex_fee_bps) - self.calculate_creator_fee(dex_fee_bps)
    }

    pub fn get_fee_breakdown(&self, dex_fee_bps: u16) -> (u16, u16, u16) {
        (self.calculate_factory_fee(dex_fee_bps), self.calculate_creator_fee(dex_fee_bps), self.calculate_lp_fee(dex_fee_bps))
    }

    pub fn validate_dex_fee(&self, fee_bps: u16) -> bool {
        fee_bps >= self.minimum_dex_fee_bps && fee_bps <= self.maximum_dex_fee_bps
    }
}

#[repr(C)]
#[derive(Debug)]
pub struct DexInstanceData {
    pub dex_id: [u8; 32],
    pub dex_name_hash: [u8; 32],
    pub owner_lock_hash: [u8; 32],
    pub dex_fee_bps: u16,
    pub factory_fee_bps: u16,
    pub creator_fee_bps: u16,
    pub lp_fee_bps: u16,
    pub pool_count: u64,
    pub total_volume: u64,
    pub total_fees_paid: u64,
    pub status: u8,
    pub bump: u64,
    pub reserved: [u8; 160],
}

impl DexInstanceData {
    pub fn new(dex_id: [u8; 32], owner_lock_hash: [u8; 32], dex_fee_bps: u16) -> Self {
        Self {
            dex_id, dex_name_hash: [0u8; 32], owner_lock_hash,
            dex_fee_bps, factory_fee_bps: 0, creator_fee_bps: 0, lp_fee_bps: 0,
            pool_count: 0, total_volume: 0, total_fees_paid: 0,
            status: 0, bump: 0, reserved: [0u8; 160],
        }
    }

    pub fn to_bytes(&self) -> [u8; DEX_INSTANCE_SIZE] {
        let mut bytes = [0u8; DEX_INSTANCE_SIZE];
        bytes[0..32].copy_from_slice(&self.dex_id);
        bytes[32..64].copy_from_slice(&self.dex_name_hash);
        bytes[64..96].copy_from_slice(&self.owner_lock_hash);
        bytes[96..98].copy_from_slice(&self.dex_fee_bps.to_le_bytes());
        bytes[104..112].copy_from_slice(&self.total_volume.to_le_bytes());
        bytes[168] = self.status;
        bytes[248..256].copy_from_slice(&self.bump.to_le_bytes());
        bytes
    }

    pub fn from_bytes(bytes: &[u8]) -> Result<Self, &'static str> {
        if bytes.len() != DEX_INSTANCE_SIZE {
            return Err("Invalid DEX instance data length");
        }
        let mut dex_id = [0u8; 32];
        let mut dex_name_hash = [0u8; 32];
        let mut owner_lock_hash = [0u8; 32];
        dex_id.copy_from_slice(&bytes[0..32]);
        dex_name_hash.copy_from_slice(&bytes[32..64]);
        owner_lock_hash.copy_from_slice(&bytes[64..96]);
        let dex_fee_bps = u16::from_le_bytes(bytes[96..98].try_into().unwrap());
        let total_volume = u64::from_le_bytes(bytes[104..112].try_into().unwrap());
        let status = bytes[168];
        let bump = u64::from_le_bytes(bytes[248..256].try_into().unwrap());
        Ok(Self {
            dex_id, dex_name_hash, owner_lock_hash, dex_fee_bps,
            factory_fee_bps: 0, creator_fee_bps: 0, lp_fee_bps: 0,
            pool_count: 0, total_volume, total_fees_paid: 0,
            status, bump, reserved: [0u8; 160],
        })
    }
}

pub fn load_factory_data() -> Result<FactoryData, &'static str> {
    let data = load_cell_data(0, Source::GroupOutput).map_err(|_| "Failed to load cell data")?;
    FactoryData::from_bytes(&data)
}

pub fn load_dex_instance_data() -> Result<DexInstanceData, &'static str> {
    let data = load_cell_data(0, Source::GroupOutput).map_err(|_| "Failed to load cell data")?;
    DexInstanceData::from_bytes(&data)
}
